library/posture.c

```diff
@@ -545,3 +585,273 @@ static void ziti_pr_handle_process(ziti_context ztx, char *id, char *path, bool
+        }
+        free(signers);
+    }
+}
+
+static int hash_sha512(uv_loop_t *loop, const char *path, unsigned char **out_buf, size_t *out_len) {
+    size_t digest_size = crypto_hash_sha512_bytes();
+    unsigned char *digest = NULL;
+    int rc = 0;
+
+#define CHECK(op) do{ rc = (op); if (rc != 0) { \
+ZITI_LOG(ERROR, "failed hashing op[" #op "]: %d", rc); \
+goto cleanup;                                   \
+} }while(0)
+
+    uv_fs_t ft;
+    uv_file file = uv_fs_open(loop, &ft, path, UV_FS_O_RDONLY, 0, NULL);
+
+    if (file < 0) { return -1; }
+    uv_buf_t buf = uv_buf_init(malloc(64 * 1024), 64 * 1024);
+    int64_t offset = 0;
+    crypto_hash_sha512_state sha512;
+    crypto_hash_sha512_init(&sha512);
+
+    // hash data
+    while (true) {
+        int read = uv_fs_read(loop, &ft, file, &buf, 1, offset, NULL);
+        if (read == 0) {
+            break;
+        }
+
+        if (read < 0) {
+            rc = -1;
+            goto cleanup;
+        }
+
+        offset += read;
+        CHECK(crypto_hash_sha512_update(&sha512, (uint8_t *) buf.base, read));
+    }
+    digest = malloc(digest_size);
+    CHECK(crypto_hash_sha512_final(&sha512, digest));
+
+    *out_buf = digest;
+    *out_len = digest_size;
+
+    cleanup:
+    if (rc != 0) FREE(digest);
+    uv_fs_close(loop, &ft, file, NULL);
+
+    return rc;
+}
+
+static bool check_running(uv_loop_t *loop, const char *path) {
+    bool result = false;
+#if _WIN32
+    HANDLE sh = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+    if (sh == NULL) {
+        ZITI_LOG(ERROR, "failed: %d", GetLastError());
+    }
+    PROCESSENTRY32W entry = {
+            .dwSize = sizeof(PROCESSENTRY32W)
+    };
+
+    char fullPath[1024];
+    DWORD fullPathSize;
+
+    for (BOOL ret = Process32FirstW(sh, &entry); ret; ret = Process32NextW(sh, &entry)) {
+        HANDLE ph = OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, entry.th32ProcessID);
+        if (ph == NULL) {
+            continue;
+        }
+        fullPathSize = sizeof(fullPath);
+        QueryFullProcessImageNameA(ph, 0, fullPath, &fullPathSize);
+
+        if (strnicmp(path, fullPath, fullPathSize) == 0) {
+            result = true;
+            break;
+        }
+    }
+    CloseHandle(sh);
+
+#elif __linux || __linux__
+
+    uv_fs_t fs_proc;
+    uv_fs_scandir(loop, &fs_proc, "/proc", 0, NULL);
+    uv_dirent_t de;
+    uv_fs_t ex;
+    char proc_path[128];
+    while (!result && uv_fs_scandir_next(&fs_proc, &de) != UV_EOF) {
+        if (de.type == UV_DIRENT_DIR) {
+            snprintf(proc_path, sizeof(proc_path), "/proc/%s/exe", de.name);
+            if (uv_fs_readlink(loop, &ex, proc_path, NULL) == 0) {
+                if (strcmp((const char *) ex.ptr, path) == 0) {
+                    result = true;
+                }
+                free(ex.ptr);
+            }
+        }
+    }
+
+#else
+    uv_utsname_t uname;
+    uv_os_uname(&uname);
+    ZITI_LOG(WARN, "not implemented on %s", uname.sysname);
```

Comment:
better log message maybe? "cannot check for process running on %s. not implemented" ?

Reply:
log message shows function name